knowledge-base/adr/0031-resilience-options.md

```diff
@@ -61,310 +61,153 @@ So we have something for circuit breaker and timeout.
-- Implementation checks if global config is present and uses them in the request
+The options are super basic - if you need them either create a feature request or implement it on your own:
 
 ```ts
-//Variant A
-function globalResilience(options: ResilienceOptions) {}
-
-//Variant B
-function globalResilience(middleWare: Middleware) {}
-
-//Variant C
-function globalResilience(middleWare: Middleware, id: string) {}
+type ResilienceOptions = {
+  retry?: boolean; // default false
+  timeouty?: boolean; // default true
+  circuitBreaker?: boolean; // default true
+};
 
-function clearGlobalResilience(id?: string) {}
+const [timeout, circuitBreaker] = resilience();
+const [timeout, circuitBreaker, retry] = resilience({ retry: true });
+const [timeout, retry] = resilience({ circuitBreaker: false, retry: true });
 ```
 
-### Parameter Passing
+### Adjusting/Replacing Provided Implementations
 
-We will extend the `HttpRequestOptions` with the resilience related options and pass these around:
+Assume a user wants to create a specific timeout for non BTP services.
+This is done the follwing way:
 
 ```ts
-function executeHttpRequest<T extends HttpRequestConfigWithOrigin>(
-  destination: DestinationOrFetchOptions,
-  requestConfig: T,
-  options?: HttpRequestOptions
-) {
-  //implementation
+const [timeout, circuitBreaker] = resilience();
+function myTimeout(inOut: MiddlewareInOut) {
+  if (inOut.context.category === 'user-defined') {
+    const wrapped = (args: any[]) =>
+      exitChain
+        ? fn(args)
+        : Promise.race([fn(args), timeoutPromise<T>(longerTimeout)]); //custom super long timeout
+    return {
+      fn: wrapped,
+      context,
+      exitChain
+    };
+  }
+  return timeout;
 }
 
-type HttpRequestOptions = {
-  csrfTokenFetching: boolean;
-  resilience?: MiddleWare;
-};
+testApi.requestBuilder().setMiddleware([myTimeout, circuitBreaker]);
 ```
 
-The resilience value could be a totally custom middleware, one with custom options or if not given the default.
-If methods execute http request they will get the additional option.
-For example the `getDestination()` or `serviceToken()` methods:
+If you want to do something completely custom you just rewrite the whole function without using the SDK implementation at all.
 
-```ts
-function getDestination(
-  options: DestinationFetchOptions & DestinationForServiceBindingOptions,
-  httpRequestOptions?: HttpRequestOptions //contains resilience
-) {
-  //implementation
-}
+### Libraries
 
-function serviceToken(
-  service: string | Service,
-  options?: OldOptions & { resilience?: MiddlewareOptions }
-) {
-  //implementation
-}
-```
+The SDK uses the following libraries to implement resilience:
+
+- For retry, we will use [async retry](https://www.npmjs.com/package/async-retry)
+- For circuit breaker, we will use [opossum](https://www.npmjs.com/package/opossum)
 
 ### Deprecation Period
 
 We will have a deprecation period for the old options.
```

Comment:
[q] If we do this totally in v3, do we still need a deprecation period? It was quite complex when supporting both setups.

Reply:
But maybe we want the migration effort to be minimal. 😄

Reply:
To deprecate something it is of course nicer to have the alternative, but perhaps we can do it because to have both in parallel is hard as you mention.